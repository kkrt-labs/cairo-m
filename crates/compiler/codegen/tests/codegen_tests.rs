//! Main test runner for CASM code generation.

use cairo_m_compiler_codegen::{CodeGenerator, CodegenDb};
use cairo_m_compiler_mir::{generate_mir, MirDb};
use cairo_m_compiler_parser::Upcast;
use cairo_m_compiler_semantic::{File, SemanticDb};
use insta::assert_snapshot;
use std::fs;
use std::path::Path;

/// Test database that implements all required traits for code generation
#[salsa::db]
#[derive(Clone, Default)]
pub struct TestDatabase {
    storage: salsa::Storage<Self>,
}

#[salsa::db]
impl salsa::Database for TestDatabase {}

#[salsa::db]
impl cairo_m_compiler_parser::Db for TestDatabase {}

#[salsa::db]
impl SemanticDb for TestDatabase {}

#[salsa::db]
impl MirDb for TestDatabase {}

#[salsa::db]
impl CodegenDb for TestDatabase {}

impl Upcast<dyn cairo_m_compiler_parser::Db> for TestDatabase {
    fn upcast(&self) -> &(dyn cairo_m_compiler_parser::Db + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn cairo_m_compiler_parser::Db + 'static) {
        self
    }
}

impl Upcast<dyn SemanticDb> for TestDatabase {
    fn upcast(&self) -> &(dyn SemanticDb + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn SemanticDb + 'static) {
        self
    }
}

impl Upcast<dyn MirDb> for TestDatabase {
    fn upcast(&self) -> &(dyn MirDb + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn MirDb + 'static) {
        self
    }
}

pub fn test_db() -> TestDatabase {
    TestDatabase::default()
}

/// The result of running code generation on a test source.
pub struct CodegenOutput {
    pub casm_code: String,
}

/// Runs the full compilation pipeline from source to CASM.
pub fn check_codegen(source: &str, path: &str) -> CodegenOutput {
    let db = test_db();
    let file = File::new(&db, source.to_string(), path.to_string());

    // Generate MIR from source
    let mir_module = generate_mir(&db, file).expect("MIR generation failed");

    let mut generator = CodeGenerator::new();
    generator
        .generate_module(&mir_module)
        .expect("CASM generation failed");

    // Generate debug representation for testing
    let casm_code = generator.debug_instructions();

    CodegenOutput { casm_code }
}

/// Loads a test case from a file.
pub fn load_test_source(path_str: &str) -> String {
    let path = Path::new("tests/test_cases").join(path_str);
    fs::read_to_string(&path).unwrap_or_else(|e| panic!("Failed to read test file {path:?}: {e}"))
}

/// A macro to define a codegen test case. It loads a source file,
/// runs the full compilation pipeline, and snapshots the CASM output.
macro_rules! codegen_test {
    ($test_name:ident, $subdir:expr) => {
        #[test]
        fn $test_name() {
            // Construct the path to the test source file.
            let path = concat!($subdir, "/", stringify!($test_name), ".cm");

            // Load the test source.
            let source = load_test_source(path);

            // Generate CASM from the source code.
            let codegen_output = check_codegen(&source, path);

            // Use insta to snapshot the entire compilation output.
            let snapshot_content = format!(
                "---\nsource: {}\nexpression: codegen_output\n---\nFixture: {}.cm\n============================================================\nSource code:\n{}\n============================================================\nGenerated CASM:\n{}\n",
                file!(),
                stringify!($test_name),
                source,
                codegen_output.casm_code,
            );
            assert_snapshot!(concat!($subdir, "_", stringify!($test_name)), snapshot_content);
        }
    };
}

// ====== Test Groups ======

// --- Simple Functions ---
codegen_test!(function_simple, "simple");
codegen_test!(function_with_params, "simple");
codegen_test!(function_with_return, "simple");

// --- Arithmetic ---
codegen_test!(add_two_numbers, "arithmetic");
codegen_test!(subtract_numbers, "arithmetic");

// --- Control Flow ---
codegen_test!(simple_if, "control_flow");
codegen_test!(if_else, "control_flow");
codegen_test!(if_else_with_merge, "control_flow");

// --- Functions ---
codegen_test!(simple_call, "functions");
codegen_test!(fib, "functions");
