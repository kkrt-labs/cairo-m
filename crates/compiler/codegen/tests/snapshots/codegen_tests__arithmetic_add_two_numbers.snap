---
source: crates/compiler/codegen/tests/codegen_tests.rs
expression: snapshot_content
---
---
source: crates/compiler/codegen/tests/codegen_tests.rs
expression: codegen_output
---
Fixture: add_two_numbers.cm
============================================================
Source code:
func add_two_numbers() -> felt {
    let a = 10;
    let b = 32;
    return a + b;
}
============================================================
Generated CASM:
add_two_numbers:
add_two_numbers:
add_two_numbers_0:
   0: 6 0 10               // Store immediate: [fp+0] = 10
   1: 6 1 32               // Store immediate: [fp+1] = 32
   2: 0 0 1 2              // [fp + 2] = [fp + 0] op [fp + 1]
   3: 4 2 -3               // Return value: [fp-3] = [fp+2]
   4: 15                   // return
